pkg/kubelet/kuberuntime/kuberuntime_image.go

```diff
@@ -111,42 +104,36 @@ func (m *kubeGenericRuntimeManager) PullImage(image kubecontainer.ImageSpec, pul
 // GetImageRef gets the ID of the image which has already been in
 // the local storage. It returns ("", nil) if the image isn't in the local storage.
 func (m *kubeGenericRuntimeManager) GetImageRef(image kubecontainer.ImageSpec) (string, error) {
+	if image.Pod == nil {
+		klog.Errorf("Pod is not set in imageSpec: %v", image)
+		return "", fmt.Errorf("POD is not set in imageSpec")
+	}
 
 	// Get the image service from the map cache for this image
-	var imageServices []internalapi.ImageManagerService
+	var imageService internalapi.ImageManagerService
 	var err error
-	// Get the image service from the map cache for this image
-	imageWithoutTag := strings.Split(image.Image, ":")[0]
-	if is := m.getPodImageService(imageWithoutTag); is != nil {
+
+	if is, err := m.GetImageServiceByPod(image.Pod); err == nil {
 		klog.Infof("Got desired image service for image %v, %v", image, is)
-		imageServices = []internalapi.ImageManagerService{is}
+		imageService = is
 	} else {
-		klog.Infof("Cannot find desired image service for image %s, loop through all image services to pull", image.Image)
-		imageServices, err = m.GetAllImageServices()
-		if err != nil {
-			klog.Errorf("GetAllImageServices Failed: %v", err)
-			return "", err
-		}
+		klog.Errorf("Cannot find desired image service for imageSpec: %v, Error: %v", image, err)
+		return "", err
 	}
 
-	for _, imageService := range imageServices {
-		status, err := imageService.ImageStatus(&runtimeapi.ImageSpec{Image: image.Image})
-		if err != nil || status == nil {
-			klog.Infof("ImageStatus for image %q failed: %v", image, err)
-		} else {
-			klog.Errorf("Got ImageStatus for image %q: %v", image, status)
-			return status.Id, nil
-		}
-
+	status, err := imageService.ImageStatus(&runtimeapi.ImageSpec{Image: image.Image})
+	if err != nil || status == nil {
+		klog.Infof("ImageStatus for image %q failed: %v", image, err)
+	} else {
+		klog.Errorf("Got ImageStatus for image %q: %v", image, status)
+		return status.Id, nil
 	}
 
 	// if not found, return "", nil
 	return "", nil
 }
 
 // ListImages gets all images currently on the machine.
-// TODO: what if the dup image from different imageService ?
```

Comment:
no code change addressing the concern recorded by the comment; is it not a real concern? If so, removing the comment makes sense.

Reply:
yes, it should be safe to remove. the Image struct uses ImageId which, in vm runtime, generated from the full image data path from the filesystem. so even with multiple runtime services, the imageID conflict situation should be skipped regardless other runtime service's logic to generated the ID of the image.